src/v/kafka/server/protocol.cc

```diff
@@ -147,16 +176,22 @@ ss::future<> protocol::apply(net::server::resources rs) {
         ? security::sasl_server::sasl_state::initial
         : security::sasl_server::sasl_state::complete);
 
+    std::optional<security::tls::mtls_state> mtls_state;
+    if (authn_method == config::broker_authn_method::mtls_identity) {
+        mtls_state = co_await get_mtls_principal_state(
+          _mtls_principal_mapper, *rs.conn);
+    }
+
     auto ctx = ss::make_lw_shared<connection_context>(
       *this,
       std::move(rs),
       std::move(sasl),
       authn_method != config::broker_authn_method::none,
-      authn_method == config::broker_authn_method::mtls_identity);
+      mtls_state);
 
-    return ss::do_until(
-             [ctx] { return ctx->is_finished_parsing(); },
-             [ctx] { return ctx->process_one_request(); })
+    co_return co_await ss::do_until(
```

Comment:
nit: i dunno how big the structures are on the stack (e.g. sasl_server, mtls_state, lw_shared_ptr, etc...) but they'll be kept alive for the entire lifetime of the connection. a few pointers is probably no big deal. i just don't recall how big the sasl_server and mtls_state structures are.

Reply:
Good point, I mostly left `sasl_server` alone, which contains a fair bit of stuff as well as the principal. `mtls_state` is just the `ss::sstring` of the principal.